prog.go

```diff
@@ -546,6 +548,52 @@ func (p *Program) Close() error {
 	return p.fd.Close()
 }
 
+// Instructions returns the 'translated' instruction stream of the program
+// after verification. For example, map accesses are made to reference their
+// kernel map IDs, not the FDs they had when the program was inserted.
+//
+// The first instruction is marked as a symbol using the Program's name.
+//
+// Available from 4.13. Requires CAP_BPF or equivalent.
+func (p *Program) Instructions() (asm.Instructions, error) {
```

Comment:
Could the method be on `ProgramInfo` instead? We could delay unmarshaling until the function is called and only hold onto `[]byte` until then.

As you point out the returned instructions contain ids, which means it's not possible to load the program into the kernel again. The function signature kind of implies that however. Some ideas:

* Return a different type `TranslatedInstructions []asm.Instruction` to document this
* Somehow encode into the instruction that its referencing an id, not an fd (per insns metadata?). Return an error when trying to marshal that.

The latter would also allow better pretty printing which seems nice. Maybe we do both?

Reply:
> Could the method be on `ProgramInfo` instead? We could delay unmarshaling until the function is called and only hold onto `[]byte` until then.

Yup, in retrospect, I should've put this onto `ProgramInfo` already, since two calls to `sys.ObjInfo` are necessary to fetch the actual instructions. The first call will populate `info.XlatedProgLen`, which will happen during `Program.Info()`, and the subsequent call to `ProgramInfo.Instructions()` will use the field to allocate an appropriately-sized buffer and fetch the insn stream.

> Return a different type `TranslatedInstructions []asm.Instruction` to document this

I would only go this route if aliasing the type is actually practically needed, e.g. if we want to implement extra methods on it. Not really the case here, I think.

> Somehow encode into the instruction that its referencing an id, not an fd (per insns metadata?). Return an error when trying to marshal that.

IMO this approach makes a type alias redundant, so eventually we could use this to do better pretty-prints. It won't practically solve the fact that this is surprising to the caller, whether an error is returned by the lib during marshaling or by the verifier. (although the lib will probably be friendlier :slightly_smiling_face:)

All in all, I think we can probably ignore this for now, as trying to re-insert these insns into the kernel will error out one way or another; reading the docstring will be necessary regardless.

Reply:
> reading the docstring will be necessary regardless.

Ok, please make this clearer in the doc then. Right now the reader needs understand the subtleties of BPF passed to the kernel to make the connection that these insns can't be loaded.

Reply:
Done, I've done some refactoring in `newProgramInfoFromFd` to coalesce fetching map IDs and binary insns when they're available. The decoding is deferred until `ProgramInfo.Instructions()`.